File: src/21-algoritmos-aproximacao/03-tsp-approximation/tsp_mst.c
/* TSP MST-based 2-aproximação */
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <float.h>

#define MAX_V 20
#define INF DBL_MAX

double dist[MAX_V][MAX_V];
bool visited[MAX_V];
int tour[MAX_V];
int tour_size = 0;

// Prim's MST
void prim_mst(int n, int parent[]) {
    double key[MAX_V];
    bool in_mst[MAX_V] = {false};
    
    for (int i = 0; i < n; i++) key[i] = INF;
    key[0] = 0;
    parent[0] = -1;
    
    for (int count = 0; count < n; count++) {
        double min = INF;
        int u = -1;
        
        for (int v = 0; v < n; v++) {
            if (!in_mst[v] && key[v] < min) {
                min = key[v];
                u = v;
            }
        }
        
        in_mst[u] = true;
        
        for (int v = 0; v < n; v++) {
            if (!in_mst[v] && dist[u][v] < key[v]) {
                parent[v] = u;
                key[v] = dist[u][v];
            }
        }
    }
}

// DFS na MST para construir tour
void dfs_tour(int u, int n, int parent[]) {
    visited[u] = true;
    tour[tour_size++] = u;
    
    for (int v = 0; v < n; v++) {
        if (parent[v] == u && !visited[v]) {
            dfs_tour(v, n, parent);
        }
    }
}

double tsp_mst_approx(int n) {
    int parent[MAX_V];
    prim_mst(n, parent);
    
    for (int i = 0; i < n; i++) visited[i] = false;
    tour_size = 0;
    
    dfs_tour(0, n, parent);
    tour[tour_size++] = 0; // Volta ao início
    
    double total = 0;
    for (int i = 0; i < tour_size - 1; i++) {
        total += dist[tour[i]][tour[i+1]];
    }
    
    return total;
}

int main() {
    printf("TSP MST 2-Aproximação\n\n");
    
    int n = 4;
    // Distâncias métricas
    double d[4][4] = {
        {0, 10, 15, 20},
        {10, 0, 35, 25},
        {15, 35, 0, 30},
        {20, 25, 30, 0}
    };
    
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            dist[i][j] = d[i][j];
    
    double cost = tsp_mst_approx(n);
    
    printf("Tour: ");
    for (int i = 0; i < tour_size; i++) printf("%d ", tour[i]);
    printf("\nCusto: %.0f\n", cost);
    printf("Garantia: ≤ 2× OPT\n");
    
    return 0;
}

File: src/21-algoritmos-aproximacao/03-tsp-approximation/tsp_mst.py
"""TSP MST 2-Aproximação"""
import heapq

def prim_mst(dist):
    n = len(dist)
    visited = [False] * n
    parent = [-1] * n
    key = [float('inf')] * n
    key[0] = 0
    
    for _ in range(n):
        u = min((k, i) for i, k in enumerate(key) if not visited[i])[1]
        visited[u] = True
        
        for v in range(n):
            if not visited[v] and dist[u][v] < key[v]:
                key[v] = dist[u][v]
                parent[v] = u
    
    return parent

def dfs_tour(u, adj, visited, tour):
    visited[u] = True
    tour.append(u)
    
    for v in adj.get(u, []):
        if not visited[v]:
            dfs_tour(v, adj, visited, tour)

def tsp_mst_approx(dist):
    n = len(dist)
    parent = prim_mst(dist)
    
    # Construir adjacências da MST
    adj = {}
    for v in range(1, n):
        u = parent[v]
        adj.setdefault(u, []).append(v)
    
    # DFS para tour
    visited = [False] * n
    tour = []
    dfs_tour(0, adj, visited, tour)
    tour.append(0)
    
    # Calcular custo
    cost = sum(dist[tour[i]][tour[i+1]] for i in range(len(tour)-1))
    
    return tour, cost

# Exemplo
dist = [
    [0, 10, 15, 20],
    [10, 0, 35, 25],
    [15, 35, 0, 30],
    [20, 25, 30, 0]
]

tour, cost = tsp_mst_approx(dist)
print("TSP MST 2-Aproximação")
print(f"Tour: {tour}")
print(f"Custo: {cost}")
print("Garantia: ≤ 2× OPT")
